govcd/user.go

```diff
@@ -0,0 +1,463 @@
+/*
+ * Copyright 2019 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+)
+
+// Definition of an OrgUser
+type OrgUser struct {
+	User     *types.User
+	client   *Client
+	adminOrg *AdminOrg // needed to be able to update, as the list of roles is found in the Org
+}
+
+// Simplified structure to insert or modify an organization user
+type OrgUserConfiguration struct {
+	Name            string // Mandatory
+	Password        string // Mandatory
+	RoleName        string // Mandatory
+	ProviderType    string // Optional: defaults to "INTEGRATED"
+	IsEnabled       bool   // Optional: defaults to false
+	IsLocked        bool   // Only used for updates
+	DeployedVmQuota int    // Optional: 0 means "unlimited"
+	StoredVmQuota   int    // Optional: 0 means "unlimited"
+	FullName        string
+	Description     string
+	EmailAddress    string
+	Telephone       string
+	TimeOut         time.Duration // Optional: default to 10 seconds
+}
+
+const (
+	// Common role names and provider types are kept here to reduce hard-coded text and prevent mistakes
+	// Roles that are added to the organization need to be entered as free text
+
+	OrgUserRoleOrganizationAdministrator = "Organization Administrator"
+	OrgUserRoleCatalogAuthor             = "Catalog Author"
+	OrgUserRoleVappAuthor                = "vApp Author"
+	OrgUserRoleVappUser                  = "vApp User"
+	OrgUserRoleConsoleAccessOnly         = "Console Access Only"
+	OrgUserRoleDeferToIdentityProvider   = "Defer to Identity Provider"
+
+	// Allowed values for provider types
+	OrgUserProviderIntegrated = "INTEGRATED" // The user is created locally or imported from LDAP
+	OrgUserProviderSAML       = "SAML"       // The user is imported from a SAML identity provider.
+	OrgUserProviderOAUTH      = "OAUTH"      // The user is imported from an OAUTH identity provider
+)
+
+// Used to check the validity of provider type on creation
+var OrgUserProviderTypes = []string{
+	OrgUserProviderIntegrated,
+	OrgUserProviderSAML,
+	OrgUserProviderOAUTH,
+}
+
+// NewUser creates an empty user
+func NewUser(cli *Client, org *AdminOrg) *OrgUser {
+	return &OrgUser{
+		User:     new(types.User),
+		client:   cli,
+		adminOrg: org,
+	}
+}
+
+// GetUserByNameOrId retrieves an user within an admin organization
+// by either name or ID
+// Returns a valid user if it exists. If it doesn't, returns nil and ErrorEntityNotFound
+func (adminOrg *AdminOrg) GetUserByNameOrId(identifier string, willRefresh bool) (*OrgUser, error) {
```

Comment:
How does ID look like in this case? Is there a corner case where someone may be using a name which conflicts would conflict with some other user's ID?

Reply:
The ID looks like `urn:vcloud:user:765a94b2-414b-4981-aff1-439687c8dd48`
and it can't conflict with a name because it can't be used as a name (colons are not allowed)